File: card-core/src/test/java/com/jean202/cardmizer/core/application/DeleteSpendingRecordServiceTest.java
package com.jean202.cardmizer.core.application;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import com.jean202.cardmizer.core.port.out.DeleteSpendingRecordPort;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import org.junit.jupiter.api.Test;

class DeleteSpendingRecordServiceTest {

    @Test
    void deleteExistingRecordDelegatesToPort() {
        CapturingDeletePort port = new CapturingDeletePort(Set.of(UUID.randomUUID()));
        DeleteSpendingRecordService service = new DeleteSpendingRecordService(port);
        UUID targetId = port.knownIds.iterator().next();

        service.delete(targetId);

        assertEquals(targetId, port.deletedId);
    }

    @Test
    void deleteNonExistentRecordThrowsResourceNotFoundException() {
        DeleteSpendingRecordPort port = id -> {
            throw new ResourceNotFoundException("Spending record not found: " + id);
        };
        DeleteSpendingRecordService service = new DeleteSpendingRecordService(port);

        UUID unknownId = UUID.randomUUID();
        ResourceNotFoundException exception = assertThrows(
                ResourceNotFoundException.class,
                () -> service.delete(unknownId)
        );
        assertEquals("Spending record not found: " + unknownId, exception.getMessage());
    }

    @Test
    void deleteAlreadyDeletedRecordThrowsResourceNotFoundException() {
        UUID recordId = UUID.randomUUID();
        Set<UUID> deletedAlready = new HashSet<>();
        deletedAlready.add(recordId);

        DeleteSpendingRecordPort port = id -> {
            if (deletedAlready.contains(id)) {
                throw new ResourceNotFoundException("Spending record not found: " + id);
            }
        };
        DeleteSpendingRecordService service = new DeleteSpendingRecordService(port);

        assertThrows(ResourceNotFoundException.class, () -> service.delete(recordId));
    }

    private static final class CapturingDeletePort implements DeleteSpendingRecordPort {
        private final Set<UUID> knownIds;
        private UUID deletedId;

        CapturingDeletePort(Set<UUID> knownIds) {
            this.knownIds = new HashSet<>(knownIds);
        }

        @Override
        public void delete(UUID id) {
            if (!knownIds.contains(id)) {
                throw new ResourceNotFoundException("Spending record not found: " + id);
            }
            this.deletedId = id;
        }
    }
}

File: card-core/src/test/kotlin/com/jean202/cardmizer/core/application/DeleteSpendingRecordServiceTest.kt
package com.jean202.cardmizer.core.application

import com.jean202.cardmizer.core.port.out.DeleteSpendingRecordPort
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import java.util.UUID

class DeleteSpendingRecordServiceTest {

    @Test
    fun deleteExistingRecordDelegatesToPort() {
        val port = CapturingDeletePort(setOf(UUID.randomUUID()))
        val service = DeleteSpendingRecordService(port)
        val targetId = port.knownIds.iterator().next()

        service.delete(targetId)

        assertEquals(targetId, port.deletedId)
    }

    @Test
    fun deleteNonExistentRecordThrowsResourceNotFoundException() {
        val port = DeleteSpendingRecordPort { id ->
            throw ResourceNotFoundException("Spending record not found: $id")
        }
        val service = DeleteSpendingRecordService(port)

        val unknownId = UUID.randomUUID()
        val exception = assertThrows(ResourceNotFoundException::class.java) {
            service.delete(unknownId)
        }
        assertEquals("Spending record not found: $unknownId", exception.message)
    }

    @Test
    fun deleteAlreadyDeletedRecordThrowsResourceNotFoundException() {
        val recordId = UUID.randomUUID()
        val deletedAlready = mutableSetOf(recordId)

        val port = DeleteSpendingRecordPort { id ->
            if (id in deletedAlready) {
                throw ResourceNotFoundException("Spending record not found: $id")
            }
        }
        val service = DeleteSpendingRecordService(port)

        assertThrows(ResourceNotFoundException::class.java) { service.delete(recordId) }
    }

    private class CapturingDeletePort(knownIds: Set<UUID>) : DeleteSpendingRecordPort {
        val knownIds: MutableSet<UUID> = HashSet(knownIds)
        var deletedId: UUID? = null

        override fun delete(id: UUID) {
            if (id !in knownIds) throw ResourceNotFoundException("Spending record not found: $id")
            deletedId = id
        }
    }
}
